File: leetcode/count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int goodNodes(TreeNode root) {
        // if a nodes val is larger than all of its preceeding nodes' values
        // then it is a good node
        if(root == null)
            return 0;
        
        Stack<TreeNode> stack = new Stack<TreeNode>();
        Stack<Integer> max = new Stack<Integer>();
        int res = 0; 
        int m = 0;
        
        stack.push(root);
        max.push(root.val);
        while(!stack.isEmpty()) {
            root = stack.pop();
            m = max.pop();
            
            if(root.val >= m) {
                // good node
                ++res; 
                m = root.val; 
            }
            
            if(root.left != null) {
                stack.push(root.left);
                max.push(m);
            }
            if(root.right != null) {
                stack.push(root.right);
                max.push(m);
            }
        }
        
        return res;
    }
}

File: leetcode/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        """
        Count the number of nodes that are good
        -A node is good if there are no nodes between the node and the root,
        where node.val > curr_node.val (including the root itself)

        Solution
        - Traverse the tree (recursively, inorder)
        - Track the largest node in the path from root to the current node, at each
            level of recursion

        """
        self.good = 0
        self.dfs(root, float('-INF'))
        return self.good


    def dfs(self, root: TreeNode, max_val: int) -> None:
        if root.val >= max_val:
            self.good += 1

        if root.left: self.dfs(root.left, max(max_val, root.val))
        if root.right: self.dfs(root.right, max(max_val, root.val))
